Create a GMRES subsection in Solver parameters

naliboff opened the pull request:
Partially addresses #2023. Moving these two parameters does not require running the update script, but moving any of the other solver parameters would require doing so.

jdannberg commented:
Keep in mind that even though we may not use any of these parameters in our tests, other people might still use them in their input file, so we should add an entry in the update script anyway. Also, I would vote for moving `Number of cheap Stokes solver steps` and `Maximum number of expensive Stokes solver steps` too. Thoughts?

naliboff commented:
I agree we should move `Number of cheap Stokes solver steps` and `Maximum number of expensive Stokes solver steps`. Beyond these two, I think we would need to move all of the rest as a group. For example, I don't think it would make sense to have the `Nonlinear solver scheme` and `Nonlinear solver tolerance` in different sections.

For these (and any other) parameters I don't think there is any need to use an additional subsection under `Solver parameters`. 

Once we decide on the remaining parameters to move, I'll modify the update script.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -338,6 +321,26 @@ namespace aspect
 
     prm.enter_subsection ("Solver parameters");
     {
+      prm.enter_subsection ("GMRES parameters");
```
Could I advocate for this to be renamed "Linear solver parameters"? The fact that we use GMRES is not really material in this regard -- it's just what we happen to use -- whereas the important fact is that it's about solving the *linear* problem.

bangerth replied:
That makes sense.

naliboff commented:
If we keep 'Linear solver tolerance' outside of the Solver parameters subsection, I think it makes sense to keep `Use direct solver for Stokes system` in the same place. However, If the group consensus is to move it, no real objection.

naliboff commented:
I've updated the subsection name to `Linear solver parameters` and moved `Number of cheap Stokes solver steps` and` Maximum number of expensive Stokes solver steps` here. 

Unless there are more parameters to move into this subsection or `Solver parameters` in general, I'll revise the update script. Many of the tests should fail before this update.

bangerth commented:
On 04/13/2018 02:37 PM, naliboff wrote:
> If we keep 'Linear solver tolerance' outside of the Solver parameters 
> subsection, I think it makes sense to keep |Use direct solver for Stokes 
> system| in the same place. However, If the group consensus is to move it, no 
> real objection.

I'd say move both. They are both clearly related to the linear solver.

naliboff commented:
@gassmoeller - The last commit addressed your two comments. Should I squash everything down to one commit before changing the update script?

gassmoeller commented:
I think I fixed the update script for all special cases we had. @naliboff, sorry for taking over your PR, I rebased everything after #2168 was merged, so you will need to throw away your local branch and check out this one if you want to do any more changes. Hopefully that should not be necessary though. I removed some parameter lines that just set parameters to their default value, and made sure that the benchmarks and cookbooks are still nice to read.
Time for a final review and merge?

jdannberg commented on cookbooks/global_no_melt.prm:
```diff
@@ -7,15 +7,22 @@ set Dimension                              = 2
 set End time                               = 1.4e8
 
+# We choose a stricter than default linear Stokes solver tolerance,
+# to improve the nonlinear convergence behavior.
```
This would also have the advantage that the files in the doc/manual/cookbooks and the cookbooks folder have the same documentation.

jdannberg commented on doc/manual/cookbooks/benchmarks/burstedde/burstedde.prm:
```diff
@@ -27,3 +26,9 @@ end
+  subsection Stokes solver parameters
+    set Number of cheap Stokes solver steps = 0
+  end
+end
```
If this was not in the original file, you don't have to add this here (in particular as there's no documentation on why the parameter is chosen like this).

jdannberg commented on doc/manual/cookbooks/benchmarks/burstedde/burstedde.prm:
```diff
@@ -1,4 +1,3 @@
-set Linear solver tolerance                = 1e-12
```
This is not the default value, so you can not just remove it. Please add it below to the Stokes solver parameters.

gassmoeller commented:
Done. @tjhei: It seems the tester is taking a break, can you have a look? Thanks!

jdannberg commented on source/simulator/parameters.cc:
```diff
@@ -269,77 +269,6 @@ namespace aspect
-                       "not converge and return an error message pointing out that the user didn't allow "
-                       "a sufficiently large number of iterations for the iterative solver to converge.");
-
     prm.declare_entry ("Temperature solver tolerance", "1e-12",
```
I thought we had agreed to also move the temperature and composition solver tolerance if we move the linear solver tolerance? 
I am not insisting on this (and the last time we talked about this, I had assumed none of these tolerances would be moved, that is why I argued they should stay where they are), but I just think this can be quite confusing. Why is the Linear solver tolerance in its own subsection, while the temperature and composition solver tolerances aren't? 
We can also do that in a follow-up pull request, but we should at least discuss again what the final state for the release should be.
